File: platform/macos/Settings/NativeRemoteResizeDraft.swift
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
import Combine
import Foundation

public enum NativeRemoteResizeSource: String, CaseIterable, Sendable {
  case custom, allDisplays, selectedDisplays
  public var title: String {
    switch self {
    case .custom: return String(localized:"settings.resize.custom.size", defaultValue:"Custom size")
    case .allDisplays: return String(localized:"settings.resize.all.local.displays", defaultValue:"All local displays")
    case .selectedDisplays: return String(localized:"settings.resize.selected.local.displays", defaultValue:"Selected local displays")
    }
  }
}

@MainActor public final class NativeRemoteResizeDraft: ObservableObject, Identifiable {
  public nonisolated let id = UUID()
  @Published public var source: NativeRemoteResizeSource = .custom
  @Published public var selectedDisplays = Set<NativeDisplayID>()
  @Published public var devicePixels = false
  @Published public private(set) var displaySnapshot: NativeDisplaySnapshot?
  private weak var displays: NativeDisplayService?
  private var reviewedDisplayGeneration: UInt64?
  @Published public var width = ""
  @Published public var height = ""
  @Published public private(set) var baseline: NativeRemoteDesktop?
  @Published public private(set) var isBusy = false
  @Published public private(set) var needsReload = false
  @Published public private(set) var message: String?
  @Published public private(set) var didApply = false
  private weak var session: NativeSession?
  private var operation: Task<Void,Never>?
  private var observations = Set<AnyCancellable>()
  private var stopped = false
  public init(session: NativeSession, displays: NativeDisplayService? = nil) {
    self.session = session; self.displays = displays
    if let displays {
      displaySnapshot = displays.snapshot
      selectedDisplays = Set(displays.snapshot.displays.map(\.id))
      displays.$snapshot.sink { [weak self] value in self?.displaySnapshot = value }.store(in:&observations)
    }
    session.$snapshot.sink { [weak self] value in MainActor.assumeIsolated {
      guard let self else { return }
      if let baseline = self.baseline, value.state != .connected || value.generation != baseline.snapshot.generation {
        self.needsReload = true; self.didApply = false
      }
      self.objectWillChange.send()
    }}.store(in:&observations)
    session.$isViewOnly.sink { [weak self] _ in self?.objectWillChange.send() }.store(in:&observations)
    session.$isClosing.sink { [weak self] value in if value { self?.stop() } }.store(in:&observations)
  }
  deinit { operation?.cancel() }
  public var canReload: Bool { !stopped && !isBusy && session?.snapshot.state == .connected && session?.isClosing == false }
  private var dimensions: (UInt32,UInt32)? {
    guard !width.isEmpty, !height.isEmpty, width.utf8.allSatisfy({ $0 >= 48 && $0 <= 57 }),
          height.utf8.allSatisfy({ $0 >= 48 && $0 <= 57 }), let w = UInt32(width), let h = UInt32(height),
          (1...65535).contains(w), (1...65535).contains(h) else { return nil }
    return (w,h)
  }
  public var needsDisplayReload: Bool {
    source != .custom && displaySnapshot?.generation != reviewedDisplayGeneration
  }
  public var chosenDisplays: [NativeDisplay] {
    guard let snapshot = displaySnapshot else { return [] }
    return source == .allDisplays ? snapshot.displays : snapshot.displays.filter { selectedDisplays.contains($0.id) }
  }
  public var missingDisplays: [NativeDisplayID] {
    guard let snapshot = displaySnapshot, source == .selectedDisplays else { return [] }
    return selectedDisplays.filter { snapshot.display($0) == nil }.sorted { $0.rawValue < $1.rawValue }
  }
  public var displayLayout: NativeDisplayLayout? {
    guard source != .custom, displays != nil, displaySnapshot?.error == nil, missingDisplays.isEmpty else { return nil }
    return try? NativeDisplayLayout(displays:chosenDisplays,devicePixels:devicePixels)
  }
  public var displayMessage: String? {
    guard source != .custom else { return nil }
    if needsDisplayReload { return String(localized:"settings.resize.local.displays.changed.reload.to.review.the.new.arrangement.before.resizing", defaultValue:"Local displays changed. Reload to review the new arrangement before resizing.") }
    if displays == nil || displaySnapshot?.error != nil { return String(localized:"settings.resize.local.display.information.is.unavailable", defaultValue:"Local display information is unavailable.") }
    if !missingDisplays.isEmpty { return String(localized:"settings.resize.a.selected.display.is.disconnected.reconnect.it.or.remove.it.from.the", defaultValue:"A selected display is disconnected. Reconnect it or remove it from the selection.") }
    if chosenDisplays.isEmpty { return String(localized:"settings.fullscreen.select.at.least.one.display", defaultValue:"Select at least one display.") }
    if displayLayout == nil { return String(localized:"settings.resize.this.arrangement.cannot.be.mapped.displays.must.not.overlap.or.mirror.and", defaultValue:"This arrangement cannot be mapped. Displays must not overlap or mirror, and the layout must fit within 65535 × 65535 pixels.") }
    return nil
  }
  private var proposedLayout: NativeRemoteLayout? {
    guard let baseline else { return nil }
    if source != .custom { return try? displayLayout?.remoteLayout(matching:baseline.layout) }
    guard let dimensions, let first = baseline.layout.screens.first else { return nil }
    return try? NativeRemoteLayout(width:dimensions.0,height:dimensions.1,
      screens:[.init(id:first.id,x:0,y:0,width:dimensions.0,height:dimensions.1,flags:first.flags)])
  }
  public var canApply: Bool {
    guard canReload, !needsReload, !needsDisplayReload, let session, let baseline, let proposedLayout,
          session.generation == baseline.snapshot.generation, session.snapshot.supportsResize,
          !session.snapshot.resizePending, !session.isViewOnly else { return false }
    return proposedLayout.width != baseline.layout.width || proposedLayout.height != baseline.layout.height ||
      proposedLayout.screens.sorted { $0.id < $1.id } != baseline.layout.screens.sorted { $0.id < $1.id }
  }
  public func reload() {
    guard canReload, let session else { return }
    do {
      displays?.refresh()
      reviewedDisplayGeneration = displaySnapshot?.generation
      let current = try session.desktopLayout(); baseline = current
      width = String(current.layout.width); height = String(current.layout.height)
      needsReload = false; didApply = false; message = nil
    } catch { needsReload = true; message = String(localized:"settings.resize.the.current.remote.desktop.size.is.unavailable", defaultValue:"The current remote desktop size is unavailable.") }
  }
  public func apply() {
    // Re-read the OS immediately before admission; notifications can be delayed.
    if source != .custom { displays?.refresh() }
    guard canApply, let session, let baseline, let layout = proposedLayout else { return }
    do {
      guard try session.desktopLayout().layout == baseline.layout else {
        needsReload = true; message = String(localized:"settings.resize.the.server.s.desktop.layout.changed.reload.before.resizing", defaultValue:"The server’s desktop layout changed. Reload before resizing."); return
      }
    } catch { message = String(localized:"settings.resize.the.requested.desktop.size.is.unavailable", defaultValue:"The requested desktop size is unavailable."); return }
    isBusy = true; message = nil; didApply = false
    operation = Task { @MainActor [weak self] in
      do {
        _ = try await session.requestDesktopLayout(layout,expectedGeneration:baseline.snapshot.generation)
        guard let self, !self.stopped else { self?.finish(); return }
        guard session.generation == baseline.snapshot.generation else { throw NativeError(.stale,"Connection changed") }
        let current = try session.desktopLayout(); self.baseline = current
        self.width = String(current.layout.width); self.height = String(current.layout.height)
        self.didApply = true; self.needsReload = false
        self.message = String(localized:"settings.resize.server.size", defaultValue:"The server’s desktop is now \((current.layout.width).formatted()) × \((current.layout.height).formatted()) pixels.")
      } catch {
        if let self, !self.stopped {
          if let failure = error as? NativeCommandFailure, failure.reason == .serverRejected {
            self.message = String(localized:"settings.resize.server.rejection", defaultValue:"The server rejected the requested size (result \((failure.nativeResult).formatted())).")
          } else if let failure = error as? NativeCommandFailure, failure.reason == .timedOut {
            self.message = String(localized:"settings.resize.the.server.has.not.replied.wait.for.its.reply.or.reconnect.before", defaultValue:"The server has not replied. Wait for its reply or reconnect before resizing again.")
          } else if let issue = error as? NativeError, issue.status == .resourceLimit {
            self.message = String(localized:"settings.resize.this.size.exceeds.the.connection.s.framebuffer.limit.choose.a.smaller.size", defaultValue:"This size exceeds the connection’s framebuffer limit. Choose a smaller size.")
          } else { self.message = String(localized:"settings.resize.the.resize.did.not.complete.reload.the.current.desktop.size.before.trying", defaultValue:"The resize did not complete. Reload the current desktop size before trying again."); self.needsReload = true }
        }
      }
      self?.finish()
    }
  }
  private func finish() { isBusy = false; operation = nil }
  public func stop() { stopped = true; observations.removeAll(); operation?.cancel() }
  public func close() async { stop(); await operation?.value }
}

File: platform/windows/TidyVNC.Native/Settings/NativeRemoteResizeDraft.cs
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
using System.ComponentModel;
using System.Globalization;
using CommunityToolkit.Mvvm.ComponentModel;
using TidyVNC.Native.Platform;

namespace TidyVNC.Native;

public enum NativeRemoteResizeSource { Custom, AllDisplays, SelectedDisplays }

public enum NativeRemoteResizeMessage
{
    Unavailable, LayoutChanged, RequestUnavailable, Applied, Rejected, TimedOut, TooLarge, Incomplete,
}

/// <summary>
/// Resize the remote desktop (macOS NativeRemoteResizeDraft; PARITY D01-D04):
/// a custom size or one remote screen per chosen local display, requested
/// against the layout it was read at; the server's reply completes it.
/// UI thread only.
/// </summary>
public sealed partial class NativeRemoteResizeDraft : ObservableObject, IDisposable
{
    private readonly NativeSession session;
    private readonly NativeDisplayService? displays;
    private ulong? reviewedDisplayGeneration;
    private CancellationTokenSource? requesting;
    private Task? operation;
    private bool stopped;

    [ObservableProperty] public partial NativeRemoteResizeSource Source { get; set; }
    [ObservableProperty] public partial IReadOnlySet<string> SelectedDisplays { get; set; } = new HashSet<string>();
    [ObservableProperty] public partial bool DevicePixels { get; set; }
    [ObservableProperty] public partial string Width { get; set; } = "";
    [ObservableProperty] public partial string Height { get; set; } = "";
    [ObservableProperty] public partial NativeRemoteDesktop? Baseline { get; private set; }
    [ObservableProperty] public partial bool IsBusy { get; private set; }
    [ObservableProperty] public partial bool NeedsReload { get; private set; }
    [ObservableProperty] public partial NativeRemoteResizeMessage? Message { get; private set; }
    /// <summary>The server's numeric result for a rejection.</summary>
    [ObservableProperty] public partial uint RejectionResult { get; private set; }
    [ObservableProperty] public partial bool DidApply { get; private set; }

    public NativeRemoteResizeDraft(NativeSession session, NativeDisplayService? displays = null)
    {
        this.session = session;
        this.displays = displays;
        if (displays is not null)
        {
            SelectedDisplays = displays.Snapshot.Displays.Select(d => d.Id).ToHashSet(StringComparer.Ordinal);
            displays.PropertyChanged += DisplaysChanged;
        }
        session.PropertyChanged += SessionChanged;
        PropertyChanged += (_, e) =>
        {
            if (e.PropertyName is nameof(Source) or nameof(SelectedDisplays) or nameof(DevicePixels) or nameof(Width) or nameof(Height)) Notify();
        };
    }

    public NativeDisplaySnapshot? DisplaySnapshot => displays?.Snapshot;

    private void DisplaysChanged(object? sender, PropertyChangedEventArgs e)
    {
        OnPropertyChanged(nameof(DisplaySnapshot));
        Notify();
    }

    private void SessionChanged(object? sender, PropertyChangedEventArgs e)
    {
        if (e.PropertyName == nameof(NativeSession.IsClosing) && session.IsClosing) { Stop(); return; }
        if (e.PropertyName == nameof(NativeSession.Snapshot) && Baseline is { } baseline &&
            (session.Snapshot.State != NativeSessionState.Connected || session.Snapshot.Generation != baseline.Snapshot.Generation))
        {
            NeedsReload = true; DidApply = false;
        }
        Notify();
    }

    private void Notify()
    {
        OnPropertyChanged(nameof(CanReload)); OnPropertyChanged(nameof(CanApply)); OnPropertyChanged(nameof(DisplayProblem));
        OnPropertyChanged(nameof(DisplayLayout)); OnPropertyChanged(nameof(NeedsDisplayReload));
    }

    public bool CanReload => !stopped && !IsBusy && session is { IsClosing: false, Snapshot.State: NativeSessionState.Connected };

    private (uint Width, uint Height)? Dimensions =>
        Width.Length != 0 && Height.Length != 0 && Width.All(char.IsAsciiDigit) && Height.All(char.IsAsciiDigit) &&
        uint.TryParse(Width, NumberStyles.None, CultureInfo.InvariantCulture, out var w) &&
        uint.TryParse(Height, NumberStyles.None, CultureInfo.InvariantCulture, out var h) && w is >= 1 and <= 65535 && h is >= 1 and <= 65535
            ? (w, h) : null;

    public bool NeedsDisplayReload => Source != NativeRemoteResizeSource.Custom && DisplaySnapshot?.Generation != reviewedDisplayGeneration;

    public IReadOnlyList<NativeDisplayInfo> ChosenDisplays => DisplaySnapshot is not { } snapshot ? []
        : Source == NativeRemoteResizeSource.AllDisplays ? snapshot.Displays : [.. snapshot.Displays.Where(d => SelectedDisplays.Contains(d.Id))];

    public IReadOnlyList<string> MissingDisplays => DisplaySnapshot is not { } snapshot || Source != NativeRemoteResizeSource.SelectedDisplays ? []
        : [.. SelectedDisplays.Where(id => snapshot.Find(id) is null).Order(StringComparer.Ordinal)];

    public NativeDisplayLayout? DisplayLayout
    {
        get
        {
            if (Source == NativeRemoteResizeSource.Custom || displays is null || DisplaySnapshot?.Error is not null || MissingDisplays.Count != 0) return null;
            try { return new NativeDisplayLayout(ChosenDisplays, DevicePixels); }
            catch (NativeError) { return null; }
        }
    }

    /// <summary>Why the chosen displays cannot be requested (macOS displayMessage).</summary>
    public NativeText? DisplayProblem
    {
        get
        {
            if (Source == NativeRemoteResizeSource.Custom) return null;
            if (NeedsDisplayReload) return new("settings.resize.local.displays.changed.reload.to.review.the.new.arrangement.before.resizing");
            if (displays is null || DisplaySnapshot?.Error is not null) return new("settings.resize.local.display.information.is.unavailable");
            if (MissingDisplays.Count != 0) return new("settings.resize.a.selected.display.is.disconnected.reconnect.it.or.remove.it.from.the");
            if (ChosenDisplays.Count == 0) return new("settings.fullscreen.select.at.least.one.display");
            if (DisplayLayout is null) return new("settings.resize.this.arrangement.cannot.be.mapped.displays.must.not.overlap.or.mirror.and");
            return null;
        }
    }

    private NativeRemoteLayout? ProposedLayout
    {
        get
        {
            if (Baseline is not { } baseline) return null;
            try
            {
                if (Source != NativeRemoteResizeSource.Custom) return DisplayLayout?.RemoteLayout(baseline.Layout);
                if (Dimensions is not { } size || baseline.Layout.Screens.Count == 0) return null;
                var first = baseline.Layout.Screens[0];
                return new NativeRemoteLayout(size.Width, size.Height, [new NativeRemoteScreen(first.Id, 0, 0, size.Width, size.Height, first.Flags)]);
            }
            catch (NativeError) { return null; }
        }
    }

    public bool CanApply
    {
        get
        {
            if (!CanReload || NeedsReload || NeedsDisplayReload || Baseline is not { } baseline || ProposedLayout is not { } proposed ||
                session.Generation != baseline.Snapshot.Generation || !session.Snapshot.SupportsResize || session.Snapshot.ResizePending || session.IsViewOnly)
                return false;
            return !proposed.Equals(baseline.Layout);
        }
    }

    public void Reload()
    {
        if (!CanReload) return;
        try
        {
            displays?.Refresh();
            reviewedDisplayGeneration = DisplaySnapshot?.Generation;
            var current = session.DesktopLayout();
            Baseline = current;
            Width = current.Layout.Width.ToString(CultureInfo.InvariantCulture);
            Height = current.Layout.Height.ToString(CultureInfo.InvariantCulture);
            NeedsReload = false; DidApply = false; Message = null;
        }
        catch (NativeError) { NeedsReload = true; Message = NativeRemoteResizeMessage.Unavailable; }
        Notify();
    }

    public void Apply()
    {
        // Read the OS again right before admission; notifications can lag.
        if (Source != NativeRemoteResizeSource.Custom) displays?.Refresh();
        if (!CanApply || Baseline is not { } baseline || ProposedLayout is not { } layout) return;
        try
        {
            if (!session.DesktopLayout().Layout.Equals(baseline.Layout))
            {
                NeedsReload = true; Message = NativeRemoteResizeMessage.LayoutChanged;
                Notify();
                return;
            }
        }
        catch (NativeError)
        {
            Message = NativeRemoteResizeMessage.RequestUnavailable;
            Notify();
            return;
        }
        IsBusy = true; Message = null; DidApply = false;
        requesting?.Dispose();
        requesting = new CancellationTokenSource();
        operation = Run(layout, baseline.Snapshot.Generation);
        Notify();
    }

    private async Task Run(NativeRemoteLayout layout, ulong generation)
    {
        try
        {
            await session.RequestDesktopLayoutAsync(layout, generation);
            if (!stopped)
            {
                if (session.Generation != generation) throw new NativeError(NativeStatus.Stale, "Connection changed");
                var current = session.DesktopLayout();
                Baseline = current;
                Width = current.Layout.Width.ToString(CultureInfo.InvariantCulture);
                Height = current.Layout.Height.ToString(CultureInfo.InvariantCulture);
                DidApply = true; NeedsReload = false;
                Message = NativeRemoteResizeMessage.Applied;
            }
        }
        catch (Exception error) when (error is NativeCommandFailure or NativeError or OperationCanceledException)
        {
            if (!stopped)
            {
                switch (error)
                {
                    case NativeCommandFailure { Reason: NativeCommandFailure.FailureReason.ServerRejected } rejected:
                        RejectionResult = rejected.NativeResult;
                        Message = NativeRemoteResizeMessage.Rejected;
                        break;
                    case NativeCommandFailure { Reason: NativeCommandFailure.FailureReason.TimedOut }:
                        Message = NativeRemoteResizeMessage.TimedOut;
                        break;
                    case NativeError { Status: NativeStatus.ResourceLimit }:
                        Message = NativeRemoteResizeMessage.TooLarge;
                        break;
                    default:
                        Message = NativeRemoteResizeMessage.Incomplete; NeedsReload = true;
                        break;
                }
            }
        }
        IsBusy = false; operation = null;
        Notify();
    }

    public void Stop()
    {
        if (stopped) return;
        stopped = true;
        session.PropertyChanged -= SessionChanged;
        if (displays is not null) displays.PropertyChanged -= DisplaysChanged;
        requesting?.Cancel();
        Notify();
    }

    /// <summary>A request already sent cannot be undone: close waits for its reply.</summary>
    public async Task CloseAsync()
    {
        Stop();
        if (operation is { } running) await running;
    }

    public void Dispose()
    {
        Stop();
        requesting?.Dispose();
    }
}
